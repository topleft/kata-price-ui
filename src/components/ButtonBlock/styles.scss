@import '../../styles/variables.scss';

.button-block {
  border-radius: 3px;
  display: block;
  width: 80%;
  margin: 0 auto;
  text-align: center;
  overflow: hidden;

  @media screen and (min-width: 540px) {
    display: block;
    width: 300px;
  }

  &__button {
    background-color: $grey-light;
    border: none;
    color: $grey-dark;
    display: block;
    padding: $space $space*3;
    text-transform: uppercase;
    width: 100%;

    @media screen and (min-width: 540px) {
      display: inline-block;
      width: 50%;
    }

    &.active {
      color: $white;
      background-color: $green;
    }

  }



}
