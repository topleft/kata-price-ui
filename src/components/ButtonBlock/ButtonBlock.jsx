import React from 'react';
import PropTypes from 'prop-types';
import './styles.scss';

class ButtonBlock extends React.PureComponent {
  constructor() {
    super();
    this.state = {
      activeIndex: 0,
    };
  }

  handleClick(activeIndex, value) {
    this.setState({activeIndex});
    this.props.handleClick(value);
  }

  render() {
    const {buttons} = this.props;
    return (
      <div className='button-block'>
        {buttons.map((button, i) => {
          const isActive = this.state.activeIndex === i ? 'active' : '';
          return <button
            key={i}
            className={`button-block__button ${isActive}`}
            onClick={() => this.handleClick(i, button.value)} >
            {button.label}
          </button>;
        })}
      </div>
    );
  }
}

ButtonBlock.propTypes = {
  buttons: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    value: PropTypes.any,
  })).isRequired,
  handleClick: PropTypes.func,
};

export default ButtonBlock;
