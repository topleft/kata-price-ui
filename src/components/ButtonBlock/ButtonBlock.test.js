import React from 'react';
import {shallow} from 'enzyme';
import ButtonBlock from './ButtonBlock';

const buttons = [
  {label: 'Monthly', value: 'mo'},
  {label: 'Annually', value: 'year'}
];

describe('ButtonBlock', () => {

  it('renders successfully when buttons are passed', () => {
    const wrapper = shallow(<ButtonBlock buttons={buttons} />);
    expect(wrapper.exists());
  });

  it('renders the correct number of buttons', () => {
    const wrapper = shallow(<ButtonBlock buttons={buttons} />);
    expect(wrapper.find('button.button-block__button').length).toBe(2);
    wrapper.setProps({buttons: [...buttons, {}]});
    expect(wrapper.find('button.button-block__button').length).toBe(3);
  });

  it('defaults the first button to active', () => {
    const wrapper = shallow(<ButtonBlock buttons={buttons}/>);
    const button1 = wrapper.find('button').first();
    expect(button1.hasClass('active')).toBe(true);
  });

  it('sets the clicked button to active', () => {
    const clickFunc = jest.fn();
    const wrapper = shallow(<ButtonBlock buttons={buttons} handleClick={clickFunc} />);
    expect(wrapper.find('button').at(0).hasClass('active')).toBe(true);
    wrapper.find('button').at(1).simulate('click');
    expect(wrapper.find('button').at(0).hasClass('active')).toBe(false);
    expect(wrapper.find('button').at(1).hasClass('active')).toBe(true);
  });

  it('calls the handleClick prop with the button.value argument', () => {
    const clickFunc = jest.fn();
    const wrapper = shallow(<ButtonBlock buttons={buttons} handleClick={clickFunc} />);
    const button0 = wrapper.find('button').at(0);
    const button1 = wrapper.find('button').at(1);
    button0.simulate('click');
    expect(clickFunc).toHaveBeenCalledWith(buttons[0].value);
    button1.simulate('click');
    expect(clickFunc).toHaveBeenCalledWith(buttons[1].value);
  });

  it('matches snapshot', () => {
    const tree = shallow(<ButtonBlock buttons={buttons}/>);
    expect(tree).toMatchSnapshot();
  });

});
