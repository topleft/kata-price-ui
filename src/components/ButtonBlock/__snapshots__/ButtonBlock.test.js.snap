// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonBlock matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ButtonBlock
    buttons={
      Array [
        Object {
          "label": "Monthly",
          "value": "mo",
        },
        Object {
          "label": "Annually",
          "value": "year",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="button-block__button active"
          onClick={[Function]}
        >
          Monthly
        </button>,
        <button
          className="button-block__button "
          onClick={[Function]}
        >
          Annually
        </button>,
      ],
      "className": "button-block",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": "Monthly",
          "className": "button-block__button active",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Monthly",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "host",
        "props": Object {
          "children": "Annually",
          "className": "button-block__button ",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Annually",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="button-block__button active"
            onClick={[Function]}
          >
            Monthly
          </button>,
          <button
            className="button-block__button "
            onClick={[Function]}
          >
            Annually
          </button>,
        ],
        "className": "button-block",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": "Monthly",
            "className": "button-block__button active",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Monthly",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": "Annually",
            "className": "button-block__button ",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Annually",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
